File: ts-nodejs/test/theme.test.ts
import { describe, expect, test } from "./bun-shim.ts";
import {
  MOONDARK,
  MOONLIT,
  effectiveTheme,
  parseLightThemeDword,
  systemThemeSync,
} from "../src/core/theme.ts";

describe("palette (SPEC 11.1)", () => {
  test("nine slots, both themes complete", () => {
    for (const palette of [MOONLIT, MOONDARK]) {
      expect(Object.keys(palette)).toHaveLength(9);
      for (const value of Object.values(palette)) expect(value).toMatch(/^#[0-9A-F]{6}$/);
    }
    expect(Object.keys(MOONLIT)).toEqual(Object.keys(MOONDARK));
  });

  test("the accent is identical in both themes and the rest are not", () => {
    expect(MOONLIT.accent).toBe("#1A88FF");
    expect(MOONDARK.accent).toBe("#1A88FF");
    expect(MOONLIT.windowBg).toBe("#F3F4F6");
    expect(MOONDARK.windowBg).toBe("#17191E");
    expect(MOONLIT.textPrimary).toBe("#1F2329");
    expect(MOONDARK.textPrimary).toBe("#F2F3F5");
    expect(MOONLIT.progressTrack).toBe("#E5E7EB");
    expect(MOONDARK.progressTrack).toBe("#3A3E47");
    expect(MOONLIT.badgeFg).toBe("#E06D00");
    expect(MOONDARK.badgeFg).toBe("#F0A040");
  });

  test("the spec-only card_bg is not part of the palette", () => {
    expect("cardBg" in MOONLIT).toBe(false);
  });
});

describe("effective theme (SPEC 11.1, 20)", () => {
  test("only the literal light/dark pin the theme; anything else follows the OS", () => {
    expect(effectiveTheme("light", () => "dark")).toBe("light");
    expect(effectiveTheme("dark", () => "light")).toBe("dark");
    expect(effectiveTheme("system", () => "dark")).toBe("dark");
    expect(effectiveTheme("garbage", () => "light")).toBe("light");
    expect(effectiveTheme("", () => "dark")).toBe("dark");
  });

  test("the OS probe is lazy: a pinned theme never calls it", () => {
    let probed = false;
    const probe = (): "light" | "dark" => {
      probed = true;
      return "dark";
    };
    expect(effectiveTheme("light", probe)).toBe("light");
    expect(effectiveTheme("dark", probe)).toBe("dark");
    expect(probed).toBe(false);
    expect(effectiveTheme("system", probe)).toBe("dark");
    expect(probed).toBe(true);
  });
});

describe("AppsUseLightTheme parsing (SPEC 20)", () => {
  const line = (value: string): string =>
    `\r\nHKEY_CURRENT_USER\\Software\\Microsoft\\Windows\\CurrentVersion\\Themes\\Personalize\r\n    AppsUseLightTheme    REG_DWORD    ${value}\r\n\r\n`;

  test("0 is dark, 1 and anything else is light", () => {
    expect(parseLightThemeDword(line("0x0"))).toBe("dark");
    expect(parseLightThemeDword(line("0x1"))).toBe("light");
    expect(parseLightThemeDword(line("0x2"))).toBe("light");
    expect(parseLightThemeDword(line("0"))).toBe("dark");
    expect(parseLightThemeDword(line("1"))).toBe("light");
  });

  test("missing or unreadable falls back to light, like unwrap_or(1)", () => {
    expect(parseLightThemeDword(null)).toBe("light");
    expect(parseLightThemeDword("")).toBe("light");
    expect(parseLightThemeDword("ERROR: the system cannot find the file specified.")).toBe("light");
    expect(parseLightThemeDword("AppsUseLightTheme    REG_SZ    yes")).toBe("light");
  });

  test("the cached reader answers without throwing", () => {
    expect(["light", "dark"]).toContain(systemThemeSync());
  });
});

File: go/internal/core/theme_test.go
package core

import "testing"

// The palette values are SPEC 11.1 brush constants — pinned literally so a
// careless edit cannot shift a color (the TS edition asserts the same bytes).
func TestPalettesMatchSpec(t *testing.T) {
	moonlit := Palette{
		Accent:        "#1A88FF",
		WindowBg:      "#F3F4F6",
		TextPrimary:   "#1F2329",
		TextSecondary: "#6B7280",
		ProgressTrack: "#E5E7EB",
		ButtonBg:      "#E9ECF0",
		ButtonHover:   "#DCE2E9",
		BadgeBg:       "#FFF0E0",
		BadgeFg:       "#E06D00",
	}
	moondark := Palette{
		Accent:        "#1A88FF",
		WindowBg:      "#17191E",
		TextPrimary:   "#F2F3F5",
		TextSecondary: "#9AA0A8",
		ProgressTrack: "#3A3E47",
		ButtonBg:      "#2C3039",
		ButtonHover:   "#3A404B",
		BadgeBg:       "#3D2E1A",
		BadgeFg:       "#F0A040",
	}
	if Moonlit != moonlit {
		t.Errorf("Moonlit = %+v", Moonlit)
	}
	if Moondark != moondark {
		t.Errorf("Moondark = %+v", Moondark)
	}
}

func TestPaletteOfOnlyDarkSelectsMoondark(t *testing.T) {
	if PaletteOf("dark") != Moondark {
		t.Error("dark selects Moondark")
	}
	for _, v := range []string{"light", "system", "", "Dark"} {
		if PaletteOf(v) != Moonlit {
			t.Errorf("%q must select Moonlit", v)
		}
	}
}

func TestEffectiveThemeRouting(t *testing.T) {
	probes := 0
	system := func() string {
		probes++
		return "dark"
	}
	if got := EffectiveTheme("light", system); got != "light" || probes != 0 {
		t.Errorf("pinned light must not probe the OS: %q probes=%d", got, probes)
	}
	if got := EffectiveTheme("dark", system); got != "dark" || probes != 0 {
		t.Errorf("pinned dark must not probe the OS: %q probes=%d", got, probes)
	}
	if got := EffectiveTheme("system", system); got != "dark" || probes != 1 {
		t.Errorf("system follows the OS: %q probes=%d", got, probes)
	}
	if got := EffectiveTheme("nonsense", system); got != "dark" || probes != 2 {
		t.Errorf("an unknown value follows the OS (Rust catch-all): %q", got)
	}
}
